File: samples/csharp_dotnetcore/81.skills-skilldialog/DialogRootBot/Dialogs/TangentDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples.DialogRootBot.Dialogs
{
    public class TangentDialog : ComponentDialog
    {
        public TangentDialog(string dialogId = nameof(TangentDialog))
            : base(dialogId)
        {
            AddDialog(new TextPrompt(nameof(TextPrompt)));
            var waterfallSteps = new WaterfallStep[]
            {
                Step1Async,
                Step2Async
            };
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), waterfallSteps));

            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> Step1Async(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var promptMessage = MessageFactory.Text("Tangent step 1 of 2", InputHints.ExpectingInput);
            return await stepContext.PromptAsync(nameof(TextPrompt), new PromptOptions { Prompt = promptMessage }, cancellationToken);
        }

        private async Task<DialogTurnResult> Step2Async(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var promptMessage = MessageFactory.Text("Tangent step 2 of 2", InputHints.ExpectingInput);
            return await stepContext.PromptAsync(nameof(TextPrompt), new PromptOptions { Prompt = promptMessage }, cancellationToken);
        }
    }
}

File: samples/javascript_nodejs/81.skills-skilldialog/dialogRootBot/dialogs/tangentDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory, InputHints } = require('botbuilder');
const { ComponentDialog, TextPrompt, WaterfallDialog } = require('botbuilder-dialogs');

const MAIN_WATERFALL_DIALOG = 'mainWaterfallDialog';

class TangentDialog extends ComponentDialog {
    constructor() {
        super(TangentDialog.name);

        this.addDialog(new TextPrompt(TextPrompt.name))
            .addDialog(new WaterfallDialog(WaterfallDialog.name, [
                this.step1.bind(this),
                this.step2.bind(this)
            ]));

        this.initialDialogId = MAIN_WATERFALL_DIALOG;
    }

    async step1(stepContext) {
        const promptMessage = MessageFactory.text('Tangent step 1 of 2', InputHints.ExpectingInput);
        return await stepContext.prompt(TextPrompt.name, {
            prompt: promptMessage
        });
    }

    async step2(stepContext) {
        const promptMessage = MessageFactory.text('Tangent step 2 of 2', InputHints.ExpectingInput);
        return await stepContext.prompt(TextPrompt.name, {
            prompt: promptMessage
        });
    }
}

module.exports.TangentDialog = TangentDialog;
